python/lib/modeldata/dmod/modeldata/data/item_domain_detector.py

```diff
@@ -0,0 +1,211 @@
+        self._num_time_steps = df.shape[0]
+        date_range = TimeRange(begin=df.iloc[0][dt_index].to_pydatetime(), end=df.iloc[-1][dt_index].to_pydatetime())
+        return DataDomain(data_format=self.get_data_format(), continuous_restrictions=[date_range],
+                          discrete_restrictions=[cat_restriction])
+
+
+# TODO: track and record hydrofabric ids and file names of what's out there so that we recognize known versions/regions
+
+
+# TODO: might need to expand things in the future ... there are other geopackage formats (e.g., older NextGen
+#  hydrofabric versions that used "id" instead of "divide_id") and maybe we need to account for that in detectors (and
+#  in formats)
+class GeoPackageHydrofabricDomainDetector(ItemDataDomainDetector):
+
+    _data_format = DataFormat.NGEN_GEOPACKAGE_HYDROFABRIC_V2
+
+    def _is_region_string_for_conus(self, region_str: Optional[str]) -> bool:
+        """
+        Whether this is a region string signifies CONUS.
+
+        Parameters
+        ----------
+        region_str: Optional[str]
+            A region string, or ``None``.
+
+        Returns
+        -------
+        bool
+            Whether there was a region string provided that indicates CONUS.
+        """
+        if not isinstance(region_str, str):
+            return False
+        else:
+            return region_str.strip().lower() == 'conus'
+
+    def detect(self, **kwargs) -> DataDomain:
+        """
+        Detect and return the data domain.
+
+        Parameters
+        ----------
+        kwargs
+            Optional kwargs applicable to the subtype, which may enhance or add to the domain detection and generation
+            capabilities, but which should not be required to produce a valid domain.
+
+        Keyword Args
+        ------------
+        version: str
+            A version string for a constraint using the ``HYDROFABRIC_VERSION`` index.
+        region: str
+            A region string for a constraint using the ``HYDROFABRIC_REGION`` index; if provided, it will be converted
+            to lower case and have any non-alphanumeric characters removed before use.
+
+        Returns
+        -------
+        DataDomain
+            The detected domain.
+
+        Raises
+        ------
+        DmodRuntimeError
+            If it was not possible to properly detect the domain.
+        """
+        # TODO: (later) probably isn't necessary to treat separately, but don't have a good way to test yet
+        if isinstance(self._item, ReadSeeker):
+            gpkg_data = self._item.read()
+            self._item.seek(0)
+        else:
+            gpkg_data = self._item
+
+        if isinstance(kwargs.get('region'), str):
+            raw_str = kwargs['region'].strip().lower()
+            if raw_str == 'conus':
+                vpu = None
+                conus = True
+            else:
+                conus = False
+                pattern = re.compile('(vpu)([-_]?)(\d+)')
```

Comment:
It looks like some VPUs have a directional suffix:

`nextgen_01.gpkg`
`nextgen_02.gpkg`
`nextgen_03N.gpkg`
`nextgen_03S.gpkg`
`nextgen_03W.gpkg`
`nextgen_04.gpkg`
`nextgen_05.gpkg`
`nextgen_06.gpkg`
`nextgen_07.gpkg`
`nextgen_08.gpkg`
`nextgen_09.gpkg`
`nextgen_10L.gpkg`
`nextgen_10U.gpkg`
`nextgen_11.gpkg`
`nextgen_12.gpkg`
`nextgen_13.gpkg`
`nextgen_14.gpkg`
`nextgen_15.gpkg`
`nextgen_16.gpkg`
`nextgen_17.gpkg`
`nextgen_18.gpkg`